expression/generator/compare_vec.go

```diff
@@ -0,0 +1,407 @@
+		case types.CompareString(buf0.GetString(i), buf1.GetString(i)) == 0:
+			i64s[i] = 1
+{{ else if eq .type.ETName "Duration" }}
+		case buf0.GetDuration(i, 0).Compare(buf1.GetDuration(i, 0)) == 0:
+			i64s[i] = 1
+{{ else if eq .type.ETName "Datetime" }}
+		case arg0[i].Compare(arg1[i]) == 0:
+			i64s[i] = 1
+{{ else if eq .type.ETName "Decimal" }}
+		case arg0[i].Compare(&arg1[i]) == 0:
+			i64s[i] = 1
+{{ end }}
+		}
+	}
+	return nil
+}
+
+func (b *builtin{{ .compare.CompareName }}{{ .type.TypeName }}Sig) vectorized() bool {
+	return true
+}
+`))
+
+const builtinCompareVecTestHeader = `import (
+	"testing"
+
+	. "github.com/pingcap/check"
+	"github.com/pingcap/parser/ast"
+	"github.com/pingcap/tidb/types"
+)
+
+var vecGeneratedBuiltinCompareCases = map[string][]vecExprBenchCase{
+`
+
+var builtinCompareVecTestFuncHeader = template.Must(template.New("").Parse(`	ast.{{ .CompareName }}: {
+`))
+
+var builtinCompareVecTestCase = template.Must(template.New("").Parse(`		{retEvalType: types.ETInt, childrenTypes: []types.EvalType{types.ET{{ .ETName }}, types.ET{{ .ETName }}}},
+`))
+
+var builtinCompareVecTestFuncTail = `	},
+`
+
+var builtinCompareVecTestTail = `}
+
+func (s *testEvaluatorSuite) TestVectorizedGeneratedBuiltinCompareEvalOneVec(c *C) {
+	testVectorizedEvalOneVec(c, vecGeneratedBuiltinCompareCases)
+}
+
+func (s *testEvaluatorSuite) TestVectorizedGeneratedBuiltinCompareFunc(c *C) {
+	testVectorizedBuiltinFunc(c, vecGeneratedBuiltinCompareCases)
+}
+
+func BenchmarkVectorizedGeneratedBuiltinCompareEvalOneVec(b *testing.B) {
+	benchmarkVectorizedEvalOneVec(b, vecGeneratedBuiltinCompareCases)
+}
+
+func BenchmarkVectorizedGeneratedBuiltinCompareFunc(b *testing.B) {
+	benchmarkVectorizedBuiltinFunc(b, vecGeneratedBuiltinCompareCases)
+}
+`
+
+type CompareContext struct {
+	// Describe the name of CompareContext(LT/LE/GT/GE/EQ/NE/NullEQ)
+	CompareName string
+	// Compare Operators
+	Operator string
+}
+
+type typeContext struct {
+	// Describe the name of "github.com/pingcap/tidb/types".ET{{ .ETName }}
+	ETName string
+	// Describe the name of "github.com/pingcap/tidb/expression".VecExpr.VecEval{{ .TypeName }}
+	// If undefined, it's same as ETName.
+	TypeName string
+	// Describe the name of "github.com/pingcap/tidb/util/chunk".*Column.Append{{ .TypeNameInColumn }},
+	// Resize{{ .TypeNameInColumn }}, Reserve{{ .TypeNameInColumn }}, Get{{ .TypeNameInColumn }} and
+	// {{ .TypeNameInColumn }}s.
+	// If undefined, it's same as TypeName.
+	TypeNameInColumn string
+	// Same as "github.com/pingcap/tidb/util/chunk".getFixedLen()
+	Fixed bool
+}
+
+var typesMap = []typeContext{
+	{ETName: "Int", TypeNameInColumn: "Int64", Fixed: true},
+	{ETName: "Real", TypeNameInColumn: "Float64", Fixed: true},
+	{ETName: "Decimal", Fixed: true},
+	{ETName: "String", Fixed: false},
+	{ETName: "Datetime", TypeName: "Time", Fixed: true},
+	{ETName: "Duration", Fixed: false},
+	{ETName: "Json", TypeName: "JSON", Fixed: false},
+}
```

Comment:
They're defined in https://github.com/pingcap/tidb/blob/master/expression/generator/helper/helper.go.

Reply:
ok